src/main/java/org/opensearch/neuralsearch/search/query/HybridCollectorManager.java

```diff
@@ -161,28 +164,83 @@ private List<ReduceableSearchResult> getSearchResults(final List<HybridSearchCol
         List topDocs = hybridSearchCollector.topDocs();
-        if (docValueFormats != null) {
-            newTopDocs = getNewTopFieldDocs(
-                getTotalHits(this.trackTotalHitsUpTo, topDocs, hybridSearchCollector.getTotalHits()),
-                topDocs,
-                sortAndFormats.sort.getSort()
-            );
-        } else {
-            newTopDocs = getNewTopDocs(getTotalHits(this.trackTotalHitsUpTo, topDocs, hybridSearchCollector.getTotalHits()), topDocs);
+        if (isSortEnabled) {
+            return getSortedTopDocsAndMaxScore(topDocs, hybridSearchCollector);
+        }
+        return getTopDocsAndMaxScore(topDocs, hybridSearchCollector);
+    }
+
+    private TopDocsAndMaxScore getSortedTopDocsAndMaxScore(List<TopFieldDocs> topDocs, HybridSearchCollector hybridSearchCollector) {
+        TopDocs sortedTopDocs = getNewTopFieldDocs(
+            getTotalHits(this.trackTotalHitsUpTo, topDocs, hybridSearchCollector.getTotalHits()),
+            topDocs,
+            sortAndFormats.sort.getSort()
+        );
+        return new TopDocsAndMaxScore(sortedTopDocs, hybridSearchCollector.getMaxScore());
+    }
+
+    private TopDocsAndMaxScore getTopDocsAndMaxScore(List<TopDocs> topDocs, HybridSearchCollector hybridSearchCollector) {
+        List<TopDocs> rescoredTopDocs = rescore(topDocs);
+        float maxScore = calculateMaxScore(rescoredTopDocs, hybridSearchCollector.getMaxScore());
+        TopDocs finalTopDocs = getNewTopDocs(
+            getTotalHits(this.trackTotalHitsUpTo, rescoredTopDocs, hybridSearchCollector.getTotalHits()),
+            rescoredTopDocs
+        );
+        return new TopDocsAndMaxScore(finalTopDocs, maxScore);
+    }
+
+    private List<TopDocs> rescore(List<TopDocs> topDocs) {
+        List<RescoreContext> rescoreContexts = searchContext.rescore();
+        boolean shouldRescore = Objects.nonNull(rescoreContexts) && !rescoreContexts.isEmpty();
+        if (!shouldRescore) {
+            return topDocs;
+        }
+        List<TopDocs> rescoredTopDocs = topDocs;
+        for (RescoreContext ctx : rescoreContexts) {
+            rescoredTopDocs = rescoredTopDocs(ctx, rescoredTopDocs);
         }
-        return new TopDocsAndMaxScore(newTopDocs, hybridSearchCollector.getMaxScore());
+        return rescoredTopDocs;
+    }
+
+    /**
+     * Rescores the top documents using the provided context. The input topDocs may be modified during this process.
+     */
+    private List<TopDocs> rescoredTopDocs(final RescoreContext ctx, final List<TopDocs> topDocs) {
+        List<TopDocs> result = new ArrayList<>(topDocs.size());
+        for (TopDocs topDoc : topDocs) {
+            try {
+                result.add(ctx.rescorer().rescore(topDoc, searchContext.searcher(), ctx));
+            } catch (IOException exception) {
+                log.error("rescore failed for hybrid query", exception);
+                throw new OpenSearchException("rescore failed", exception);
```

Comment:
do you want to put more information in this exception message on why it failed.

Reply:
ack, agree on adding specific exception type. I'm not sure what we can do on adding more information on why error happened, the IOException is part of the interface method signature down to 3 levels on callees. I think the exception itself will have most of the info about the error